### PR TITLE
NFT governance

### PR DESCRIPTION
**Pull request by wraeth-eth**
## What does this PR do and why?

Adds the ability to:
- Add on-chain governance flags to NFT collection
- Add reserved rate to each NFT
- Add owner of reserved minted NFT
- Add governance weight of NFT

Also updated formatting of the NFT view

closes https://github.com/jbx-protocol/juice-interface/issues/2857

## Screenshots or screen recordings

Screen Shot 2023-01-23 at 11 17 45 am
Screen Shot 2023-01-23 at 11 18 34 am
Screen Shot 2023-01-23 at 11 20 07 am

## Acceptance checklist

- [ ] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [ ] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).
- [ ] I have tested this PR in dark mode and light mode (if applicable).

### CODE REVIEW

**Comment by wraeth-eth** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2812** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/2812" target="_blank">Graphite</a> 
    * **PR #2746** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/2746" target="_blank">Graphite</a> 
      * **PR #2750** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/2750" target="_blank">Graphite</a> 
        * **PR #2813** <a href="https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/2813" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/2813?utm_source=stack-comment).

**Comment by tomquirk** [comment]
@johnnyd-eth can you please take over the review here?

**Comment by johnnyd-eth** [comment]
@wraeth-eth would it be possible to give a summary and recordings demonstrating the changes you've made here?

**Comment by wraeth-eth** [comment]
@johnnyd-eth I have attached screenshots. lmk if you need more. I can take you through it in a call if you want

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 1]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+            )}
+            <Form.Item
+              name="votingWeight"
+              label={t`Voting weight`}
+              tooltip={t`Give this NFT a voting weight to be used for on-chain governance. The number you set is only used in relation to other NFTs in this collection.`}
+              extra={
+                <Trans>
+                  If you use the default governance option (no governance), the
+                  voting weight will still be accessible on the blockchain for
+                  use in snapshot strategies or any other desired purpose.
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:359]: Capital S in Snapshot

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 2]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+                  ]}
+                >
+                  <EthAddressInput />
+                </Form.Item>
+              </>
+            )}
+            <Form.Item
+              name="votingWeight"
+              label={t`Voting weight`}
+              tooltip={t`Give this NFT a voting weight to be used for on-chain governance. The number you set is only used in relation to other NFTs in this collection.`}
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:354]: Wdyt about switching the `extra` and `tooltip` texts here? I feel the tooltip is more essential, and having the current `extra` text exposed adds a lot of (in many cases, unnecessary) text to the modal.

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 3]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+                  stringOkay: true,
+                }),
+              ]}
+            >
+              <FormattedNumberInput placeholder="300" />
+            </Form.Item>
+            <Form.Item
+              name="externalUrl"
+              label={t`External link`}
+              tooltip={t`Link minters of this NFT to your project's website, discord etc.`}
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:374]: Capital D in Discord

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 4]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
-          <PrefixedInput prefix="https://" />
-        </Form.Item>
+          <CreateCollapse.Panel
+            header={<OptionalHeader header={t`Advanced options`} />}
+            key={0}
+            hideDivider
+          >
+            <Form.Item
+              extra={
+                <span className="text-sm">
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:296]: Should we make this `xs` like all the other extras on the form?

File: src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx [hunk 5]
```diff
@@ -94,6 +119,36 @@ export const NftRewardsPage = () => {
...
+                  <Radio.Group className="flex flex-col gap-5">
+                    <RadioItem
+                      value={JB721GovernanceType.NONE}
+                      title={
+                        <>
+                          <Trans>No on-chain governance capabilities</Trans>{' '}
+                          <CreateBadge.Default />
+                        </>
+                      }
+                      description={t`Your project's NFTs will not have any governance capabilities.`}
```
Comment (johnnyd-eth) [inline src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:135]: I think these should all be `xs` too.

File: src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx [hunk 6]
```diff
@@ -94,6 +119,36 @@ export const NftRewardsPage = () => {
...
+                header={<OptionalHeader header={t`On-chain Governance`} />}
+                hideDivider
+              >
+                <Form.Item name="onChainGovernance">
+                  <Radio.Group className="flex flex-col gap-5">
+                    <RadioItem
+                      value={JB721GovernanceType.NONE}
+                      title={
+                        <>
+                          <Trans>No on-chain governance capabilities</Trans>{' '}
```
Comment (johnnyd-eth) [inline src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:131]: Perhaps we could scrap 'capabilities'? Seems a bit redundant to me

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 7]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+                  <Trans>Set aside a percentage of freshly minted NFTs.</Trans>
+                </span>
+              }
+            >
+              <JuiceSwitch
+                value={isReservingNfts}
+                onChange={setIsReservingNfts}
+                label={
+                  <span className="flex items-center gap-2">
+                    <Trans>Reserve NFTs</Trans>
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:306]: Should this be 'Reserved NFTs' do you think? I'm thinking mainly to make it clear it's a similar concept to 'Reserved tokens'

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 8]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+                </Trans>
+              }
+              rules={[
+                inputIsIntegerRule({
+                  label: t`Voting weight`,
+                  stringOkay: true,
+                }),
+              ]}
+            >
+              <FormattedNumberInput placeholder="300" />
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:369]: I think we agreed to scrap placeholders totally a while ago, especially in cases of number values like this. 

However, I do think we need more of an explanation as to what value belongs here. Weight of what? 

I think after we add more info about this, there'll be enough text to warrant outsourcing most of this field's text to a docs page (not sure if one already exists). I think in the end just an extra with...
 - `Give this NFT a voting weight to be used for on-chain governance. <b>Learn more</b>.`
 
... will be all that is necessary. And then move everything else into the 'Learn more' docs page. Wdyt?

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 3 above)
Comment (johnnyd-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:374]: And 'Discord, etc.' (comma)

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 9]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+                    >
+                      <FormattedNumberInput min={2} placeholder="0" />
+                    </Form.Item>
+                  </div>{' '}
+                  <Trans>NFTs minted for:</Trans>
+                </span>
+                <Form.Item
+                  className="mb-8"
+                  name="beneficiary"
+                  label={t`Ethereum wallet address`}
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:340]: Screen Shot 2023-01-24 at 2 26 40 pm

Error message isn't going away for me here after typing in an ENS

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx [hunk 10]
```diff
@@ -236,19 +282,108 @@ export const AddEditRewardModal = ({
...
+                      className="mb-0"
+                      name="nftReservedRate"
+                      rules={[
+                        inputMustExistRule({ label: t`Reserved rate` }),
+                        inputIsIntegerRule({
+                          stringOkay: true,
+                        }),
+                      ]}
+                    >
+                      <FormattedNumberInput min={2} placeholder="0" />
```
Comment (johnnyd-eth) [inline src/components/Create/components/RewardsList/AddEditRewardModal.tsx:332]: Screen Shot 2023-01-24 at 2 39 54 pm

Error state is a bit off here. Maybe just 'Required'? And would be great to have the input field not jump up like it is but not blocking

File: src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx [hunk 11]
```diff
@@ -94,6 +119,36 @@ export const NftRewardsPage = () => {
             <CreateCollapse>
```
Comment (johnnyd-eth) [inline src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:119]: Not a huge issue, but would be great if opening this section didn't cause the whole section to expand slightly:

https://user-images.githubusercontent.com/96150256/214214050-6bf9a26a-f630-4005-ac4e-a67312e8a6fc.mp4

**Comment by johnnyd-eth** [comment]
So so good @wraeth-eth excellent work. First round of comments are in 🛩️

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 2 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:354]: image
Switching has about the same amount of text. I am basing off @strath-m initial design here:
https://www.figma.com/file/EN7CZTekcSo3sHyaGt6cC0/JBX---Project-Creation-Flow---High-Fidelity---Updated-Jan-20223?node-id=2466%3A156448&t=woX39OFlKH1vNxrF-0

The copy changed a bit in discord

File: src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx (hunk 5 above)
Comment (wraeth-eth) [reply src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:135]: These were based off @strath-m design:
image


For comparison, this is what xs looks like:
image

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 8 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:369]: Yep, strong agree with this one

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 9 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:340]: These are are quirks (and annoyances) of antd and the way that the address input code works. Probably out of scope as these are larger problems in our code :(

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 10 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:332]: We can definitely change to just required. The jump up is default behavior of antd unfortunately :(

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 9 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:340]: Actually, it looks like the problem you're facing might be different re invalid name even after entering. I'll see if I can recreate

Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:340]: Are you entering it on goerli? johnnyd.eth isn't registered on goerli so this is expected behaviour

File: src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx (hunk 11 above)
Comment (wraeth-eth) [reply src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:119]: There's something causing the divs to not fully size to 600px it seems. Will open an issue to look into, but on first look I can't seem to find it

Comment (wraeth-eth) [reply src/components/Create/components/pages/NftRewards/NftRewardsPage.tsx:119]: https://github.com/jbx-protocol/juice-interface/issues/2885

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 8 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:369]: I think we should have a greater discussion about this too. Unsure if it is blocking of this PR as we can make the change once we port over to a docs link too

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 2 above)
Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx]: Have switched them around!

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 9 above)
Comment (johnnyd-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:340]: Ahh yes good 

> Are you entering it on goerli? johnnyd.eth isn't registered on goerli so this is expected behaviour

Ahh yes good point sir, would be that.

File: src/components/Create/components/RewardsList/AddEditRewardModal.tsx (hunk 8 above)
Comment (johnnyd-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:369]: Created issue for the docs page: https://github.com/jbx-protocol/juice-interface/issues/2900

But I think really think we should scrap the placeholder, it doesn't really make sense to have here regardless.

Comment (wraeth-eth) [reply src/components/Create/components/RewardsList/AddEditRewardModal.tsx:369]: Ah yep - willr emove placeholder

**Comment by wraeth-eth** [comment]
⏳ This pull request is set to merge as part of a Graphite merge job
Stack job ID: IrWWvSudyZVz93dICWu0.
See details on [graphite.dev](https://app.graphite.dev/github/pr/jbx-protocol/juice-interface/2813)

**Pull request merged**